Mathlib/Tactic/SolveByElim.lean

```diff
@@ -65,121 +320,197 @@ This would make it impossible to apply the lemma
+
+open Syntax
 
-/-- Elaborate the context and the list of lemmas -/
-def elabContextLemmas (lemmas : List (TermElabM Expr)) (ctx : TermElabM (List Expr)) :
-    MetaM (List Expr) := Elab.Term.TermElabM.run' do
-  pure ((← lemmas.mapM id) ++ (← ctx))
-
-/-- Attempt to solve the given metavariable by repeating applying a list of facts. -/
-def solveByElimAux (lemmas : List (TermElabM Expr)) (ctx : TermElabM (List Expr)) (n : Nat) :
-    TacticM Unit := Tactic.done <|> match n with
-      | 0 => throwError "solve_by_elim exceeded its recursion limit"
-      | n + 1 => do
-  let goal ← getMainGoal
-  withTraceNode `Meta.Tactic.solveByElim
-    -- Note: the `addMessageContextFull` is so that we show the goal using the mvar context before
-    -- the `do` block below runs, potentially unifying mvars in the goal.
-    (return m!"{exceptEmoji ·} working on: {← addMessageContextFull goal}")
-    do
-      let es ← elabContextLemmas lemmas ctx
-
-      -- We attempt to find an expression which can be applied,
-      -- and for which all resulting sub-goals can be discharged using `solveByElim n`.
-      es.firstM fun e => withTraceNode `Meta.Tactic.solveByElim
-          (return m!"{exceptEmoji ·} tried to apply: {e}") do
-        liftMetaTactic (fun mvarId => mvarId.apply e)
-        solveByElimAux lemmas ctx n
-
-/-- Attempt to solve the given metavariable by repeating applying one of the given expressions,
-or a local hypothesis. -/
-def solveByElimImpl (only : Bool) (es : List (TSyntax `term)) (n : Nat) (g : MVarId) :
-    MetaM Unit := g.withContext do
-  let ⟨lemmas, ctx⟩ ← mkAssumptionSet only es
-  let _ ← Elab.Term.TermElabM.run' (Elab.Tactic.run g (solveByElimAux lemmas ctx n))
-  pure ()
+/--
+Parse the lemma argument of a call to `solve_by_elim`.
+The first component should be true if `*` appears at least once.
+The second component should contain each term `t`in the arguments.
+The third component should contain `t` for each `-t` in the arguments.
+-/
+def parseArgs (s : Option (TSyntax ``args)) :
+    Bool × List Term × List Term :=
+  let args : Array (TSyntax ``arg) := match s with
+  | some s => match s with
+    | `(args| [$args,*]) => args.getElems
+    | _ => #[]
+  | none => #[]
+  let args : Array (Option (Term ⊕ Term)) := args.map fun t => match t with
+    | `(arg| $_:star) => none
+    | `(arg| - $t:term) => some (Sum.inr t)
+    | `(arg| $t:term) => some (Sum.inl t)
+    | _ => panic! "Unreachable parse of solve_by_elim arguments."
+  let args := args.toList
+  (args.contains none,
+    args.filterMap fun o => o.bind Sum.getLeft,
+    args.filterMap fun o => o.bind Sum.getRight)
 
 /--
 `solve_by_elim` calls `apply` on the main goal to find an assumption whose head matches
 and then repeatedly calls `apply` on the generated subgoals until no subgoals remain,
-performing at most `max_depth` (currently hard-coded to 6) recursive steps.
+performing at most `maxDepth` (currently hard-coded to 12) recursive steps.
```

Comment:
No longer hardcoded, because it now accepts a `Config`, right? (Also, 6 is the default after your recent commit.)